src/az_ulib_ipc/az_ulib_ipc.c

```diff
@@ -0,0 +1,352 @@
+// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license.
+// See LICENSE file in the project root for full license information.
+
+#include <stdint.h>
+
+#include "azure_macro_utils/macro_utils.h"
+#include "umock_c/umock_c_prod.h"
+
+#include "az_pal_os_api.h"
+#include "az_ulib_action_api.h"
+#include "az_ulib_descriptor_api.h"
+#include "az_ulib_ipc_api.h"
+#include "az_ulib_base.h"
+#include "internal/az_ulib_ipc.h"
+#include "ucontract.h"
+#include "ulib_config.h"
+#include "ulib_port.h"
+#include "ulib_result.h"
+#include "ulog.h"
+
+/*
+ * IPC is a singleton component, and shall be initialized only once.
+ */
+static _az_ulib_ipc *ipc = NULL;
+
+static _az_ulib_ipc_interface *get_interface_by_name_and_version(
+    const char *const name,
+    az_ulib_version version,
+    AZ_ULIB_VERSION_MATCH_CRITERIA match_criteria) {
+  _az_ulib_ipc_interface *result = NULL;
+
+  for (size_t i = 0; i < AZ_ULIB_CONFIG_MAX_IPC_INTERFACE; i++) {
+    if ((ipc->interface_list[i].interface_descriptor != NULL)
```

Comment:
>ipc->interface_list[i].interface_descriptor != NULL [](start = 9, length = 51)

should this be a `break` criteria? if you set the max IPC interface to say 16 but only have 4 added, it would be a waste to iterate through the remaining 8 slots.

Reply:
now that i think about it it probably isn't since you could unpublish one and have a gap between valid interfaces

---
In reply to: [338314427](https://github.com/Azure/azure-ulib-c/pull/40#discussion_r338314427) [](ancestors = 338314427)